### PR TITLE
issue #13 - added create/get functions in rest api

### LINKED ISSUE #13: #28 / Setup RESTful routes for working with pastes
For this issue we're looking for an initial API design. The endpoints should be defined and return 501 Not Implemented status.

### PR DESCRIPTION
**Pull request by macskay**
Added RESTful API

Resolves #13

### CODE REVIEW

File: npm-debug.log [hunk 1]
```diff
@@ -0,0 +1,24 @@
+0 info it worked if it ends with ok
```
Comment (ghost) [inline npm-debug.log:1]: Probably need to add *.log to .gitignore

File: pasteit/public/javascripts/pasted.js [hunk 2]
```diff
@@ -18,6 +18,8 @@ $(document).ready(function() {
     editor.getDoc().setValue($("#pasted-content").val());
     editor.refresh();
 
-    $("#delete").prop("href", "/delete" + window.location.pathname);
```
Comment (ghost) [inline pasteit/public/javascripts/pasted.js]: We need to find or write a variadic join function that takes string sequences and builds a proper URI path from it. Similar to os.path.join from Python. I'm sure something like this exists.

File: pasteit/routes/api.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+router.get("/api", function(req, res) {
+    res.json({ message: 'Hooray! Welcome to the RESTful-API' });
+});
+
+router.route("/api/create")
+    .post(function(req, res) {
+        var hash = fn.newPaste(req);
+        
+        res.json({ paste_url : req.get("host") + "/p/"+hash });
```
Comment (ghost) [inline pasteit/routes/api.js]: spaces between operators.

+ "/p/" + hash

File: pasteit/routes/api.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    .post(function(req, res) {
+        var hash = fn.newPaste(req);
+        
+        res.json({ paste_url : req.get("host") + "/p/"+hash });    
+});
+
+router.route("/api/p/")
+    .post(function(req, res) {
+        var result = fn.getPaste(req.params.hash);
+
```
Comment (ghost) [inline pasteit/routes/api.js]: Consistent white space usage. Need to get a linter plugged into your dev environment.

File: pasteit/routes/api.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+        res.json({ error : "Invalid hash" }) ? result == null : res.json({ result : paste });
```
Comment (ghost) [inline pasteit/routes/api.js]: No space between the key and the value in object definitions. No space between the closing and ending brackets in inline object declarations. Use triple equals for strict type comparison of null.

This whole expression is confusing. At first glance it seems to do nothing, until you read the documentation for res.json. https://expressjs.com/en/4x/api.html#res.json. Consider the following in terms of readability.

``` javascript
 // res.json({ error : "Invalid hash" }) ? result === null : res.json({ result : paste });

  if (result == null) {
    res.status(404).send("Not found"); // if the hash didn't exist then the URL is invalid, return a 404
  } else {
    res.json({"result": paste});
  }
```

File: pasteit/routes/functions.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
+var redis = require("redis");
+var client = redis.createClient();
+var uuid = require("uuid4");
+
+
+module.exports = {
+    newPaste : function(req) {
+        var newPaste = module.exports.createPaste(req);
+        var hash = uuid();
```
Comment (ghost) [inline pasteit/routes/functions.js]: Use only one "var" per function if possible, just put a comma after (req), and put hash on the next line.

File: pasteit/routes/functions.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            "language" : req.body.language,
+            "paste" : req.body.paste,
+            "mime" : req.body.mime
+        }
+
+        return newPaste;
+    },
+
+    addNewPaste : function(hash, newPaste) {
+        client.hmset(hash, newPaste);
```
Comment (ghost) [inline pasteit/routes/functions.js]: client is a bad name for this variable. It should be "redisClient" so it's clear we're inserting into redis here.

File: pasteit/routes/functions.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    },
+
+    getPaste : function(hash, callback) {
+        client.hgetall(hash, function(err, reply) {
+            var paste = {};
+            if(reply == null)
+                paste = { error : err }
```
Comment (ghost) [inline pasteit/routes/functions.js]: Returning redis errors back to the client is not useful. If redis had a problem, the response to the API client should be a "500: Server error". Also, we should be logging any errors returned by redis to our express server.

**Review by ghost — CHANGES_REQUESTED** [review]
This was a pretty big review. My comments are broken into two buckets. Readability and behavior. We should really work hard to use consistent formatting and think more about variable names. It's for the benefit of our future selves. As far as behavior, get familiar with the different HTTP Status Codes and consider carefully what you should be returning to the client. Not finding an entity is a 404. Failing to connect to the DB or the DB returning an error is a 500 Server error. The client can not take any reasonable action to correct the issue so returning the error messages is not helpful.

Finally, REST is not RPC. The only kinds of error responses we should be sending back to the client are when they are PUT or POSTing and submit invalid data.

File: pasteit/routes/api.js [hunk 9]
```diff
@@ -9,7 +10,7 @@ router.get('/api', function(req, res) {
 router.route('/api/create')
     .post(function(req, res) {
         let hash = fn.newPaste(req);
-        res.json({paste_url: req.get('host') + '/p/' + hash});
```
Comment (ghost) [inline pasteit/routes/api.js]: Do we still need to put the slashes in if we're using join?

**Review by ghost — COMMENTED** [review]
I think you might be able to get rid of the slashes in the join function. Otherwise looks great! Merge it!

**Review by ghost — APPROVED** [review]

File: pasteit/routes/api.js (hunk 9 above)
Comment (macskay) [reply pasteit/routes/api.js]: https://stackoverflow.com/questions/9756567/do-you-need-to-use-path-join-in-node-js 2nd answer states, that you can leave them out, I guess there is some kind of sanitizing function in the end.
Changed it, tested it, working!

**Pull request merged**